### DISCUSSION TITLE
Anyone else moving to Bun + Elysia?

### DISCUSSION THREAD

**Discussion by uncomplexity** [uNetworking/uWebSockets.js#1139] [General]
I've been rawdogging uWebSockets.js with my own wrapper for years now, albeit numerous existing attempts to replicate the express api.

- There's https://github.com/nanoexpress/nanoexpress by @dalisoft 
- There's https://github.com/kartikk221/hyper-express by @kartikk221

Lately the end-to-end type-safety movement has been gaining traction, so I started exploring.

- For https://trpc.io/ there's https://github.com/romanzy313/trpc-uwebsockets by @romanzy313
- For https://ts-rest.com/ there is no uws plugin yet.
- For https://www.zodios.org/ there is no uws plugin yet.

Then I recall that https://bun.sh/ uses uws as a built-in underneath. I have tried the 0.x versions a long time ago but figured it's now time to revisit it. Installed the current version and it ran smoothly.

Then I tried https://elysiajs.com/ just today, I am floored by how well written and well documented it is.

- has end-to-end type-safety.
- has built-in type-checks using sinclair's typebox, which is better than [zod](https://x.com/sebmck/status/1787188422433497188).
- has built-in plugins for cors, swagger docs, jwt, static, etc.

Has anyone also tried it? What are your experiences?

**Comment by romanzy313** [comment]
Elysia is nice inteed! The new ts server frameworks use all the newest typescript features, which makes them very typesafe. But beaware that compile and runtime performance can degrade significantly on larger projects when the Elysia Eden rpc is used. For Elysia, I didn't really like the `set.headers.key = "value"` and I find middleware definition cumbersome. 

Another awesome modern framework is [hono](https://github.com/honojs/hono) and it runs on virtually any runtime. I think it's the most simple and straightforward. It also has end-to-end typesafety system called [stacks](https://hono.dev/docs/concepts/stacks) and I like its api a lot. Incoming request validation is really powerful and it uses zod, instead of elysias typebox. It's definately a good choice for edge environments such as cloudflare workers!

**Comment by uNetworkingAB** [comment]
The natural path would be for Node.js to go die in a corner while Deno, Bun takes over. But this assumes rationality. And in a sphere where Express and Socket.IO are still most used by far, I would guess there is no rationality. I mean look at Python. Python is very slow but grows in popularity day by day. Even PyPy isn't being used.

People in this sphere will continue to use inferior software, and write tutorials for inferior software for a looooooooong time.

**Comment by uNetworkingAB** [comment]
I can add that, I think both Deno and Bun are excellent steps forward from Node.js. Node.js is misdesigned and mismanaged and both Deno and Bun are faster in the ways uWS tries to improve Node.js.

Bun is, like mentioned, even built using uWS and uSockets so if that project takes off I would not care to see uWS.js become irrelevant.

My personal favorite right now (for simple things I sometimes build) is Deno. I've seen a lot of segfaults in Bun but Deno never fails.
